extensions/exchanges/gdax/exchange.js

```diff
@@ -1,20 +1,57 @@
 var Gdax = require('gdax')
-  , path = require('path')
-  , colors = require('colors')
-  , numbro = require('numbro')
 
 module.exports = function container (get, set, clear) {
   var c = get('conf')
 
-  var public_client = {}, authed_client
+  var public_client = {}, authed_client, websocket_client = {}, websocket_cache = {}
 
   function publicClient (product_id) {
     if (!public_client[product_id]) {
+      websocketClient(product_id)
       public_client[product_id] = new Gdax.PublicClient(product_id, c.gdax.apiURI)
     }
     return public_client[product_id]
   }
 
+  function websocketClient (product_id) {
```

Comment:
The gdax websocket api reportedly barfs if too many connections are opened at once; instead they recommend opening one connection and subscribing to multiple channels.  Unfortunately, the node api doesn't support subscribing after connection yet.

I propose submitting a pull request to the gdax node project adding a feature that allows subscription after connection (assuming the api server allows such behavior) and subscribing to all products at once in the first connection in the interim.  I'm interesting in this feature and may begin working on that pull request if I can find the time.

Reply:
Just to follow up, I added this pull request at https://github.com/coinbase/gdax-node/pull/179 such that some day it may be used.

Reply:
Subscribing to all the products at once in the websocket is going to be a firehose worth of information coming down, with the vast majority of it being irrelevant to the use case at hand. The OrderbookSync will be doing a lot of work to maintain all the orderbook status' for each product when it doesn't need to, and the trade cache will need to filter out any matches that aren't what we're really looking for or else it'll continually push irrelevant trades into that array. My presumption is that you're running multiple instances of zenbot against different products. In that case, even if you do subscribe to multiple products, a single instance of zenbot can only trade against a single product and they can't share information across instances so I'm not sure how it'd help that use case. The rate limit for websocket connections is 1/IP address/4 seconds. If you wait 4 seconds between starting zenbot instances, I _think_ you'll be ok.

Reply:
You've got a point.  My hope is actually different -- I want to use zenbot to archive the orderbook data, so I plan to connect to multiple exchanges at once.  But others shouldn't have to bear the processing for that.  Maybe if I'm lucky my patch will be merged.

I sent you a pull request so auth works at https://github.com/krystophv/zenbot/pull/1 .  It works with my credentials now.